src/main/java/uk/gov/hmcts/divorce/noticeofchange/service/ChangeOfRepresentativeService.java

```diff
@@ -0,0 +1,145 @@
+
+import jakarta.servlet.http.HttpServletRequest;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.ccd.sdk.type.ListValue;
+import uk.gov.hmcts.ccd.sdk.type.Organisation;
+import uk.gov.hmcts.divorce.divorcecase.model.ApplicationType;
+import uk.gov.hmcts.divorce.divorcecase.model.CaseData;
+import uk.gov.hmcts.divorce.divorcecase.model.Solicitor;
+import uk.gov.hmcts.divorce.idam.IdamService;
+import uk.gov.hmcts.divorce.noticeofchange.model.ChangeOfRepresentative;
+import uk.gov.hmcts.divorce.noticeofchange.model.Representative;
+import uk.gov.hmcts.divorce.solicitor.client.organisation.FindUsersByOrganisationResponse;
+import uk.gov.hmcts.divorce.solicitor.client.organisation.OrganisationClient;
+import uk.gov.hmcts.divorce.solicitor.client.organisation.ProfessionalUser;
+import uk.gov.hmcts.reform.authorisation.generators.AuthTokenGenerator;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+
+import static org.apache.http.HttpHeaders.AUTHORIZATION;
+import static uk.gov.hmcts.divorce.divorcecase.model.ApplicationType.SOLE_APPLICATION;
+import static uk.gov.hmcts.divorce.noticeofchange.model.ChangeOfRepresentationAuthor.SOLICITOR_NOTICE_OF_CHANGE;
+
+@Component
+@RequiredArgsConstructor
+public class ChangeOfRepresentativeService {
+
+    private final IdamService idamService;
+    private final HttpServletRequest request;
+    private final OrganisationClient organisationClient;
+    private final AuthTokenGenerator authTokenGenerator;
+
+    public void buildChangeOfRepresentative(CaseData caseData,
+                                            CaseData beforeCaseData,
+                                            String updatedVia,
+                                            boolean isApplicant1) {
+        Representative addedRepresentative;
+        Representative removedRepresentative = null;
+        String clientName = isApplicant1 ? caseData.getApplicant1().getFullName() : caseData.getApplicant2().getFullName();
+        String updatedBy;
+        var currentSolicitor = isApplicant1 ? caseData.getApplicant1().getSolicitor()  : caseData.getApplicant2().getSolicitor();
+
+        if (SOLICITOR_NOTICE_OF_CHANGE.getValue().equals(updatedVia)) {
+            String sysUserToken = idamService.retrieveSystemUpdateUserDetails().getAuthToken();
+            String s2sToken = authTokenGenerator.generate();
+            var changeOrganisationRequest = caseData.getChangeOrganisationRequestField();
+            var orgToAdd = changeOrganisationRequest.getOrganisationToAdd();
+            var organisationId = orgToAdd.getOrganisationId();
+            var loggedInUserEmail = changeOrganisationRequest.getCreatedBy().toLowerCase();
+            final var nocRequestingUser = getProfessionalUsers(sysUserToken, s2sToken, organisationId, loggedInUserEmail);
```

Comment:
Lots of var usage - could we specify types?  Might be overkill.

Reply:
Replaced var with actual types